frontend/src/api/actions/init.js

```diff
@@ -0,0 +1,170 @@
+import { fetchApplications } from "./applications";
+import { fetchAssignments } from "./assignments";
+import { fetchContractTemplates } from "./contract_templates";
+import { fetchInstructors } from "./instructors";
+import { fetchPositions } from "./positions";
+import { setGlobals, globalsSelector } from "./globals";
+
+/**
+ * A helper function to replace all API actions to
+ * use mock API actions.
+ *
+ * @param {boolean} enableMockAPI
+ */
+function toggleMockApi(enableMockAPI) {
+    // in production, ToggleMockApi is a no-op. In development, it actually
+    // does something.
+    /* eslint-disable */
+    if (enableMockAPI != null && process.env.REACT_APP_DEV_FEATURES) {
+        const mock = require("../mockAPI");
+        const mockAPI = mock.mockAPI;
+
+        if (enableMockAPI === true) {
+            mockAPI.replaceGlobalFetch();
+        } else {
+            mockAPI.restoreGlobalFetch();
+        }
+    }
+    /* eslint-enable */
+}
+
+/**
+ * A helper function to update all necessary global variables
+ *
+ * @param {bool} enableMockAPI
+ * @param {object} newActiveSession
+ * @param {function} getState
+ * @returns
+ */
+function updateGlobals(enableMockAPI, newActiveSession, getState) {
+    const state = getState();
+    let globals = globalsSelector(state);
+
+    if (enableMockAPI != null) {
+        globals = {
+            ...globals,
+            mockAPI: enableMockAPI
+        };
+    }
+
+    if (newActiveSession) {
+        globals = {
+            ...globals,
+            activeSession: newActiveSession.id
+        };
+    }
+
+    return setGlobals(globals);
+}
+
+/**
+ * The old way this is used to work is that after component mounted
+ * actions that needs to be made asynchronously are executed in
+ * `componentDidMount` of `App.js`, but leaving the `toggleMockAPI`
+ * controlled by the mockAPI component.
+ * However, this leads to a racing condition such that when mockAPI
+ * should be turned on, the toggling action does not occur before all
+ * the API calls are made. Hence the app is trying to fetch from the real
+ * API instead of the mock API.
+ * In addition, on the event of turning on and off the mock API, switching
+ * active user role and setting active sessions, a similar consecutive API
+ * actions are made comparing to the initialization.
+ *
+ * The new approach is to make each actions in `iniOrder` array, which
+ * contains all the consecutive API actions that need to be made on
+ * application load, as a stage, and move the actual toggling of the mock
+ * API to this level instead of leaving it to the mockAPI component. Whenever
+ * a related events occur (for example, changing active session), the action
+ * creator will call this `initFromStage` action with the matching stage and
+ * this function will perform different API actions that is necessary
+ * after the specified `stage`
+ *
+ * @export
+ * @param {string} stage
+ * @param {object} options
+ * @returns
```

Comment:
Returns a promise? Explain what it returns. And, should this function be async?

Reply:
The function returns an async function similar to `validatedApiDispatcher`. It is used this way because we need the `dispatch` and `getState` and I have used `validataedApiDispatcher` as an example.

Reply:
Great. Put that in the `@returns` part of the docstring